client/session.go

```diff
@@ -702,6 +713,94 @@ func (s *session) Truncate(namespace string) (int64, error) {
 	return truncated, resultErr.FinalError()
 }
 
+func (s *session) getPeersForShard(shard uint32) ([]hostQueue, error) {
+	s.RLock()
+	peers := make([]hostQueue, 0, s.topoMap.Replicas())
+	err := s.topoMap.RouteShardForEach(shard, func(idx int, host topology.Host) {
+		if s.origin != nil && s.origin.ID() == host.ID() {
+			// Don't include the origin host
+			return
+		}
+		peers = append(peers, s.queues[idx])
+	})
+	s.RUnlock()
+	if err != nil {
+		return nil, err
+	}
+	return peers, nil
+}
+
+type metadataIter struct {
+	inputCh  <-chan blocksMetadata
+	errCh    <-chan error
+	host     topology.Host
+	blocks   []block.Metadata
+	metadata block.BlocksMetadata
+	done     bool
+	err      error
+}
+
+func newMetadataIter(inputCh <-chan blocksMetadata, errCh <-chan error) PeerBlocksMetadataIter {
+	return &metadataIter{
+		inputCh: inputCh,
+		errCh:   errCh,
+		blocks:  make([]block.Metadata, 0, 64),
+	}
+}
+
+func (it *metadataIter) Next() bool {
+	if it.done || it.err != nil {
+		return false
+	}
+	m, more := <-it.inputCh
+	if !more {
+		it.err = <-it.errCh
+		it.done = true
+		return false
+	}
+	it.host = m.peer.Host()
+	it.blocks = it.blocks[:0]
+	for _, b := range m.blocks {
+		bm := block.NewMetadata(b.start, b.size, b.checksum)
+		it.blocks = append(it.blocks, bm)
+	}
+	it.metadata = block.NewBlocksMetadata(m.id, it.blocks)
+	return true
+}
+
+func (it *metadataIter) Current() (topology.Host, block.BlocksMetadata) {
```

Comment:
`blocksMetadata` has a `peer` field, which is the host queue where the request was served from